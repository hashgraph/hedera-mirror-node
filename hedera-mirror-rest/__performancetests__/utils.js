#!/usr/bin/env node
/*-
 *
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const fs = require('fs');
const yaml = require('js-yaml');

// Load config
process.env.CONFIG_NAME = 'perfTestConfig';
process.env.CONFIG_PATH = __dirname;
const config = require('../config');

const mustLoadYaml = fileName => {
  try {
    console.log(`Loading yaml file ${fileName}`);
    return yaml.safeLoad(fs.readFileSync(fileName, 'utf8'));
  } catch (err) {
    console.log(`Failed to load yaml file ${fileName}: ${err}`);
    process.exit(1);
  }
};

module.exports = {
  config: config,
  mustLoadYaml: mustLoadYaml
};
