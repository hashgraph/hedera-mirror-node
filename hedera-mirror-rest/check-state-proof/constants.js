/*-
 *
 * Hedera Mirror Node
 *  ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *  ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

const MAX_RECORD_LENGTH = 64 * 1024;

const MAX_TRANSACTION_LENGTH = 64 * 1024;

const SHA_384_LENGTH = 48;

const SHA_384_WITH_RSA = {
  type: 1,
  maxLength: 384,
};

const SIMPLE_SUM = 101;

module.exports = {
  MAX_RECORD_LENGTH,
  MAX_TRANSACTION_LENGTH,
  SHA_384_LENGTH,
  SHA_384_WITH_RSA,
  SIMPLE_SUM,
};
