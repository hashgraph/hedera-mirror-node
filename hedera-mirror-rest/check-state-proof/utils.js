/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */
'uses strict';

// external libraries
const fs = require('fs');
const fetch = require('node-fetch');
const AbortController = require('abort-controller');

const base64StringToBuffer = (base64String) => {
  return Buffer.from(base64String, 'base64');
};

const replaceSpecialCharsWithUnderScores = (stringToFormat) => {
  return stringToFormat.replace(/[.@\-]/g, '_');
};

const makeStateProofDir = (transactionId, stateProofJson) => {
  const newDirPath = replaceSpecialCharsWithUnderScores(transactionId);
  fs.mkdirSync(newDirPath, {recursive: true});
  fs.writeFileSync(`${newDirPath}/apiResponse.json`, JSON.stringify(stateProofJson));
  console.log(`Supporting files and API response for the state proof will be stored in the directory ${newDirPath}`);
  return newDirPath;
};

const storeFile = (data, file, ext) => {
  const newFilePath = `${replaceSpecialCharsWithUnderScores(file)}.${ext}`;
  fs.writeFileSync(`${newFilePath}`, data, (err) => {
    if (err) throw err;
  });
};

const getAPIResponse = async (url) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  console.log(`Requesting state proof files from ${url}...`);
  return await fetch(url, {signal: controller.signal})
    .then(async (response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .catch((error) => {
      throw `Error fetching ${url}: ${error}`;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
};

const readJSONFile = (filePath) => {
  const rawData = fs.readFileSync(filePath);
  return JSON.parse(rawData);
};

module.exports = {
  base64StringToBuffer,
  getAPIResponse,
  makeStateProofDir,
  readJSONFile,
  storeFile,
  replaceSpecialCharsWithUnderScores,
};
