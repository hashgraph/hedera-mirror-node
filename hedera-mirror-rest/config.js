/*-
 * ‌
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const extend = require('extend');
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');
const {InvalidConfigError} = require('./errors/invalidConfigError');
const {cloudProviders, networks, defaultBucketNames} = require('./constants');

let configName = 'application';
if (process.env.CONFIG_NAME) {
  configName = process.env.CONFIG_NAME;
}

const config = {};
let loaded = false;

function load(configPath) {
  if (!configPath) {
    return;
  }

  let configFile = path.join(configPath, `${configName}.yml`);
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }

  configFile = path.join(configPath, configName + '.yaml');
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }
}

function loadYaml(configFile) {
  try {
    const doc = yaml.safeLoad(fs.readFileSync(configFile, 'utf8'));
    console.log(`Loaded configuration source: ${configFile}`);
    extend(true, config, doc);
  } catch (err) {
    console.log(`Skipping configuration ${configFile}: ${err}`);
  }
}

function loadEnvironment() {
  for (const [key, value] of Object.entries(process.env)) {
    setConfigValue(key, value);
  }
}

/*
 * Sets a config property from an environment variable by converting HEDERA_MIRROR_REST_FOO_BAR to an object path
 * notation hedera.mirror.rest.foo.bar using a case insensitive search. If more than one property matches with a
 * different case, it will choose the first.
 */
function setConfigValue(propertyPath, value) {
  let current = config;
  let properties = propertyPath.toLowerCase().split('_');

  // Ignore properties that don't start with HEDERA_MIRROR_REST
  if (properties.length < 4 || properties[0] !== 'hedera' || properties[1] !== 'mirror' || properties[2] !== 'rest') {
    return;
  }

  for (let i in properties) {
    let property = properties[i];
    let found = false;

    for (let [k, v] of Object.entries(current)) {
      if (property === k.toLowerCase()) {
        if (i < properties.length - 1) {
          current = v;
          found = true;
          break;
        } else {
          current[k] = convertType(value);
          console.log(`Override config with environment variable ${propertyPath}=${value}`);
          return;
        }
      }
    }

    if (!found) {
      return;
    }
  }
}

function convertType(value) {
  let parsedValue = value;

  if (value !== null && value !== '' && !isNaN(value)) {
    parsedValue = +value;
  } else if (value === 'true' || value === 'false') {
    parsedValue = value === 'true';
  }

  return parsedValue;
}

function getConfig() {
  return config.hedera && config.hedera.mirror ? config.hedera.mirror.rest : config;
}

function parseStateProofStreamsConfig() {
  const {stateproof} = getConfig();
  if (!stateproof || !stateproof.enabled) {
    return;
  }

  const {streams: streamsConfig} = stateproof;
  if (!Object.values(networks).includes(streamsConfig.network)) {
    throw new InvalidConfigError(`unknown network ${streamsConfig.network}`);
  }

  if (!streamsConfig.bucketName) {
    streamsConfig.bucketName = defaultBucketNames[streamsConfig.network];
  }

  if (!streamsConfig.bucketName) {
    // the default for network 'OTHER' is null, throw err if it's not configured
    throw new InvalidConfigError('stateproof.streams.bucketName must be set');
  }

  if (!Object.values(cloudProviders).includes(streamsConfig.cloudProvider)) {
    throw new InvalidConfigError(`unsupported object storage service provider ${streamsConfig.cloudProvider}`);
  }
}

if (!loaded) {
  load(path.join(__dirname, 'config'));
  load(__dirname);
  load(process.env.CONFIG_PATH);
  loadEnvironment();
  parseStateProofStreamsConfig();
  loaded = true;
}

module.exports = getConfig();
