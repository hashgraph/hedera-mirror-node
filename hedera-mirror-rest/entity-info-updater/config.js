/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const fs = require('fs');
const log4js = require('log4js');
const path = require('path');
const yaml = require('js-yaml');

const logger = log4js.getLogger();
let config = {};
let loaded = false;

const loadYaml = () => {
  try {
    const configFile = path.join(__dirname, 'config', `application.yml`);
    logger.info(`Loaded configuration source: ${configFile}`);
    config = yaml.load(fs.readFileSync(configFile, 'utf-8'));
  } catch (err) {
    logger.info(err);
  }
};

if (!loaded) {
  loadYaml();
  loaded = true;
}

const getConfig = () => {
  return config.hedera && config.hedera.mirror ? config.hedera.mirror.entityUpdate : config;
};

module.exports = getConfig();
