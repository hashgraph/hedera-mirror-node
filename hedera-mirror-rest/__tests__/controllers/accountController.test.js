/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {getResponseLimit} from '../../config';
import * as constants from '../../constants';
import * as utils from '../../utils';

const {default: defaultLimit} = getResponseLimit();

const {
  filterKeys: {SERIAL_NUMBER, SPENDER_ID, TOKEN_ID, ORDER, LIMIT, TIMESTAMP},
} = constants;
const {
  opsMap: {eq, gt, gte, lt, lte, ne},
} = utils;

import Bound from '../../controllers/bound';
import {AccountController} from '../../controllers';

const ownerAccountId = BigInt(1);

const serialEqFilter = {key: SERIAL_NUMBER, operator: eq, value: 5};
const serialGtFilter = {key: SERIAL_NUMBER, operator: gt, value: 6};
const serialGteFilter = {key: SERIAL_NUMBER, operator: gte, value: 7};
const serialLtFilter = {key: SERIAL_NUMBER, operator: lt, value: 15};
const serialLteFilter = {key: SERIAL_NUMBER, operator: lte, value: 16};

const spenderEqFilter = {key: SPENDER_ID, operator: eq, value: 1000};
const spenderGtFilter = {key: SPENDER_ID, operator: gt, value: 2000};
const spenderGteFilter = {key: SPENDER_ID, operator: gte, value: 3000};
const spenderLtFilter = {key: SPENDER_ID, operator: lt, value: 4000};
const spenderLteFilter = {key: SPENDER_ID, operator: lte, value: 5000};
const spenderEqInFilter = {key: SPENDER_ID, operator: eq, value: 1050};
const spenderEqInFilter2 = {key: SPENDER_ID, operator: eq, value: 1100};

const tokenIdEqFilter = {key: TOKEN_ID, operator: eq, value: 100};
const tokenIdGtFilter = {key: TOKEN_ID, operator: gt, value: 101};
const tokenIdGteFilter = {key: TOKEN_ID, operator: gte, value: 102};
const tokenIdLtFilter = {key: TOKEN_ID, operator: lt, value: 150};
const tokenIdLteFilter = {key: TOKEN_ID, operator: lte, value: 151};

// only used by staking rewards tests
const timestampEqFilter = {key: TIMESTAMP, operator: eq, value: 1000};
const timestampGtFilter = {key: TIMESTAMP, operator: gt, value: 2000};
const timestampGteFilter = {key: TIMESTAMP, operator: gte, value: 3000};
const timestampLtFilter = {key: TIMESTAMP, operator: lt, value: 4000};
const timestampLteFilter = {key: TIMESTAMP, operator: lte, value: 5000};
const timestampNeFilter = {key: TIMESTAMP, operator: ne, value: 6000};
const timestampEqFilter2 = {key: TIMESTAMP, operator: eq, value: 2000};
const timestampEqFilter3 = {key: TIMESTAMP, operator: eq, value: 3000};

describe('extractNftMultiUnionQuery', () => {
  const defaultExpected = {
    bounds: {
      primary: new Bound(TOKEN_ID, 'token_id'),
      secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
    },
    lower: [],
    inner: [],
    upper: [],
    ownerAccountId,
    order: constants.orderFilterValues.DESC,
    limit: defaultLimit,
    spenderIdFilters: [],
    spenderIdInFilters: [],
  };

  const specs = [
    {
      name: 'empty',
      filters: [],
      expected: defaultExpected,
    },
    {
      name: 'order asc',
      filters: [{key: ORDER, operator: eq, value: 'asc'}],
      expected: {...defaultExpected, order: 'asc'},
    },
    {
      name: 'limit',
      filters: [{key: LIMIT, operator: eq, value: 60}],
      expected: {...defaultExpected, limit: 60},
    },
    {
      name: 'token eq',
      filters: [tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdEqFilter],
      },
    },
    {
      name: 'token gt',
      filters: [tokenIdGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: tokenIdGtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdGtFilter],
      },
    },
    {
      name: 'token lt',
      filters: [tokenIdLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: tokenIdLtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdLtFilter],
      },
    },
    {
      name: 'token gte',
      filters: [tokenIdGteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: tokenIdGteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdGteFilter],
      },
    },
    {
      name: 'token lte',
      filters: [tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: tokenIdLteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdLteFilter],
      },
    },
    {
      name: 'token lt and gt',
      filters: [tokenIdLtFilter, tokenIdGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: tokenIdGtFilter,
            upper: tokenIdLtFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdGtFilter, tokenIdLtFilter],
      },
    },
    {
      name: 'token eq and serial eq',
      filters: [serialEqFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: Bound.create({equal: serialEqFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
        },
        lower: [tokenIdEqFilter, serialEqFilter],
      },
    },
    {
      name: 'token gt and serial eq',
      filters: [serialEqFilter, tokenIdGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({equal: serialEqFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
          primary: Bound.create({lower: tokenIdGtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [tokenIdGtFilter, serialEqFilter],
      },
    },
    {
      name: 'token lt and serial eq',
      filters: [serialEqFilter, tokenIdLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({equal: serialEqFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
          primary: Bound.create({upper: tokenIdLtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [tokenIdLtFilter, serialEqFilter],
      },
    },
    {
      name: 'token gte and lte, serial eq',
      filters: [serialEqFilter, tokenIdGteFilter, tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({equal: serialEqFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
          primary: Bound.create({
            lower: tokenIdGteFilter,
            upper: tokenIdLteFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
        },
        lower: [tokenIdGteFilter, tokenIdLteFilter, serialEqFilter],
      },
    },
    {
      name: 'token gte and serial gte',
      filters: [serialGteFilter, tokenIdGteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({lower: serialGteFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
          primary: Bound.create({lower: tokenIdGteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [{...tokenIdGteFilter, operator: eq}, serialGteFilter],
        inner: [{...tokenIdGtFilter, value: tokenIdGtFilter.value + 1}],
      },
    },
    {
      name: 'token lte and serial lte',
      filters: [serialLteFilter, tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({upper: serialLteFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
          primary: Bound.create({upper: tokenIdLteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        inner: [{...tokenIdLteFilter, operator: lt}],
        upper: [{...tokenIdLteFilter, operator: eq}, serialLteFilter],
      },
    },
    {
      name: 'token gte and lte, serial gte and lte',
      filters: [serialGteFilter, serialLteFilter, tokenIdGteFilter, tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({
            lower: serialGteFilter,
            upper: serialLteFilter,
            filterKey: SERIAL_NUMBER,
            viewModelKey: 'serial_number',
          }),
          primary: Bound.create({
            lower: tokenIdGteFilter,
            upper: tokenIdLteFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
        },
        lower: [{...tokenIdGteFilter, operator: eq}, serialGteFilter],
        inner: [
          {...tokenIdGteFilter, operator: gt},
          {...tokenIdLteFilter, operator: lt},
        ],
        upper: [{...tokenIdLteFilter, operator: eq}, serialLteFilter],
      },
    },
    {
      name: 'token eq and serial lt',
      filters: [serialLtFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: Bound.create({upper: serialLtFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
        },
        lower: [tokenIdEqFilter, serialLtFilter],
      },
    },
    {
      name: 'token eq and serial gt',
      filters: [serialGtFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: Bound.create({lower: serialGtFilter, filterKey: SERIAL_NUMBER, viewModelKey: 'serial_number'}),
        },
        lower: [tokenIdEqFilter, serialGtFilter],
      },
    },
    {
      name: 'spender eq le gte',
      filters: [spenderEqFilter, spenderEqInFilter, spenderEqInFilter2, spenderLtFilter, spenderGteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: new Bound(TOKEN_ID, 'token_id'),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        spenderIdFilters: [spenderLtFilter, spenderGteFilter],
        spenderIdInFilters: [spenderEqFilter, spenderEqInFilter, spenderEqInFilter2],
      },
    },
    {
      name: 'token eq spender eq le gte',
      filters: [
        tokenIdEqFilter,
        spenderEqFilter,
        spenderEqInFilter,
        spenderEqInFilter2,
        spenderLtFilter,
        spenderGteFilter,
      ],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
          secondary: new Bound(SERIAL_NUMBER, 'serial_number'),
        },
        lower: [tokenIdEqFilter],
        spenderIdFilters: [spenderLtFilter, spenderGteFilter],
        spenderIdInFilters: [spenderEqFilter, spenderEqInFilter, spenderEqInFilter2],
      },
    },
    {
      name: 'token gte and lte, serial gte and lte, spender eq lte gt',
      filters: [
        serialGteFilter,
        serialLteFilter,
        tokenIdGteFilter,
        tokenIdLteFilter,
        spenderEqFilter,
        spenderEqInFilter,
        spenderLteFilter,
        spenderGtFilter,
      ],
      expected: {
        ...defaultExpected,
        bounds: {
          secondary: Bound.create({
            lower: serialGteFilter,
            upper: serialLteFilter,
            filterKey: SERIAL_NUMBER,
            viewModelKey: 'serial_number',
          }),
          primary: Bound.create({
            lower: tokenIdGteFilter,
            upper: tokenIdLteFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
        },
        lower: [{...tokenIdGteFilter, operator: eq}, serialGteFilter],
        inner: [
          {...tokenIdGteFilter, operator: gt},
          {...tokenIdLteFilter, operator: lt},
        ],
        upper: [{...tokenIdLteFilter, operator: eq}, serialLteFilter],
        spenderIdFilters: [spenderLteFilter, spenderGtFilter],
        spenderIdInFilters: [spenderEqFilter, spenderEqInFilter],
      },
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(AccountController.extractNftMultiUnionQuery(spec.filters, ownerAccountId)).toEqual(spec.expected);
    });
  });
});

describe('extractTokenMultiUnionQuery throw', () => {
  const specs = [
    {
      name: 'no token.id',
      filters: [{key: SERIAL_NUMBER, operator: eq, value: 1}],
    },
    {
      name: 'serial number ne',
      filters: [{key: SERIAL_NUMBER, operator: ne, value: 1}],
    },
    {
      name: 'serial number range and equal',
      filters: [serialEqFilter, serialGtFilter, serialLtFilter],
    },
    {
      name: 'serial number multiple equal',
      filters: [serialEqFilter, serialEqFilter],
    },
    {
      name: 'serial number multiple lower bounds',
      filters: [serialGtFilter, serialGteFilter],
    },
    {
      name: 'serial number multiple upper bounds',
      filters: [serialLtFilter, serialLteFilter],
    },
    {
      name: 'token.id ne',
      filters: [{key: TOKEN_ID, operator: ne, value: 1}],
    },
    {
      name: 'token.id multiple equal',
      filters: [tokenIdEqFilter, tokenIdEqFilter],
    },
    {
      name: 'token.id multiple lower bounds',
      filters: [tokenIdGtFilter, tokenIdGteFilter],
    },
    {
      name: 'token.id multiple upper bounds',
      filters: [tokenIdLtFilter, tokenIdLteFilter],
    },
    {
      name: 'token gt and serial gte',
      filters: [serialGteFilter, tokenIdGtFilter],
    },
    {
      name: 'token lt and serial lte',
      filters: [serialLteFilter, tokenIdLtFilter],
    },
    {
      name: 'spender le lte',
      filters: [tokenIdEqFilter, spenderLtFilter, spenderLteFilter],
    },
    {
      name: 'spender ge gte',
      filters: [tokenIdEqFilter, spenderGtFilter, spenderGteFilter],
    },
    {
      name: 'spender ne',
      filters: [{key: SPENDER_ID, operator: ne, value: 1}],
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(() =>
        AccountController.extractNftMultiUnionQuery(spec.filters, ownerAccountId)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('extractStakingRewardsQuery', () => {
  const defaultExpected = {
    order: constants.orderFilterValues.DESC,
    limit: defaultLimit,
    conditions: [],
    params: [],
  };

  // the params in the following specs are just the "dynamic" parts of the query; the function
  // "listStakingReardsByAccountId" in AccountController inserts the accountId at $1 and the limit at $2,
  // at which point the conditions reflect the proper parameters.
  const specs = [
    {
      name: 'empty',
      filters: [],
      expected: defaultExpected,
    },
    {
      name: 'order asc',
      filters: [{key: ORDER, operator: eq, value: 'asc'}],
      expected: {...defaultExpected, order: 'asc'},
    },
    {
      name: 'limit',
      filters: [{key: LIMIT, operator: eq, value: 60}],
      expected: {...defaultExpected, limit: 60},
    },
    {
      name: 'timestamp eq',
      filters: [timestampEqFilter],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp in ($3)'],
        params: [1000],
      },
    },
    {
      name: 'timestamp gt',
      filters: [timestampGtFilter],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp > $3'],
        params: [2000],
      },
    },
    {
      name: 'timestamp gte',
      filters: [timestampGteFilter],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp >= $3'],
        params: [3000],
      },
    },
    {
      name: 'timestamp lt',
      filters: [timestampLtFilter],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp < $3'],
        params: [4000],
      },
    },
    {
      name: 'timestamp lte',
      filters: [timestampLteFilter],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp <= $3'],
        params: [5000],
      },
    },
    {
      name: 'timestamp eq 1 and 2',
      filters: [timestampEqFilter, timestampEqFilter2],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp in ($3,$4)'],
        params: [1000, 2000],
      },
    },
    {
      name: 'timestamp eq 1 and 2 and 3',
      filters: [timestampEqFilter, timestampEqFilter2, timestampEqFilter3],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp in ($3,$4,$5)'],
        params: [1000, 2000, 3000],
      },
    },
    {
      name: 'all params',
      filters: [
        timestampGteFilter,
        timestampLteFilter,
        {key: LIMIT, operator: eq, value: 60},
        {key: ORDER, operator: eq, value: 'asc'},
      ],
      expected: {
        ...defaultExpected,
        conditions: ['srt.consensus_timestamp >= $3', 'srt.consensus_timestamp <= $4'],
        params: [3000, 5000],
        order: 'asc',
        limit: 60,
      },
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(AccountController.extractStakingRewardsQuery(spec.filters, ownerAccountId)).toEqual(spec.expected);
    });
  });
});

describe('extractStakingRewardsQuery throw', () => {
  const specs = [
    {
      name: 'timestamp ne',
      filters: [timestampNeFilter],
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(() =>
        AccountController.extractStakingRewardsQuery(spec.filters, ownerAccountId)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
