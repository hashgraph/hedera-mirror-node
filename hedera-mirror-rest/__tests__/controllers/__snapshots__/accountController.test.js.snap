// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTokenMultiUnionQuery throw no token.id 1`] = `"serialnumber without a token.id parameter filter"`;

exports[`extractTokenMultiUnionQuery throw serial number multiple equal 1`] = `"Only one equal (eq) operator is allowed for serialnumber"`;

exports[`extractTokenMultiUnionQuery throw serial number multiple lower bounds 1`] = `"Only one gt/gte operator is allowed for serialnumber"`;

exports[`extractTokenMultiUnionQuery throw serial number multiple upper bounds 1`] = `"Only one lt/lte operator is allowed for serialnumber"`;

exports[`extractTokenMultiUnionQuery throw serial number ne 1`] = `"Not equal (ne) operator is not supported for serialnumber"`;

exports[`extractTokenMultiUnionQuery throw serial number range and equal 1`] = `"Can't support both range and equal for serialnumber"`;

exports[`extractTokenMultiUnionQuery throw spender ge gte 1`] = `"Multiple range params not allowed for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender le lte 1`] = `"Multiple range params not allowed for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender ne 1`] = `"Not equals (ne) comparison operator is not supported"`;

exports[`extractTokenMultiUnionQuery throw token gt and serial gte 1`] = `"token.id must have gte or eq operator"`;

exports[`extractTokenMultiUnionQuery throw token lt and serial lte 1`] = `"token.id must have lte or eq operator"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple equal 1`] = `"Only one equal (eq) operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple lower bounds 1`] = `"Only one gt/gte operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple upper bounds 1`] = `"Only one lt/lte operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id ne 1`] = `"Not equal (ne) operator is not supported for token.id"`;

exports[`extractStakingRewardsQuery throw timestamp ne 1`] = `"Not equals (ne) operator is not supported for timestamp"`;
